polkadot/xcm/xcm-executor/src/traits/controller.rs

```diff
@@ -0,0 +1,206 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A set of traits that define how a pallet interface with XCM.
+
+use crate::{
+	traits::{QueryHandler, QueryResponseStatus},
+	InteriorMultiLocation,
+	Junctions::Here,
+	MultiLocation, Xcm, Outcome, XcmError
+};
+use frame_support::{pallet_prelude::*, parameter_types};
+use sp_weights::Weight;
+use xcm::{v3::XcmHash, VersionedMultiLocation, VersionedXcm};
+
+/// Umbrella trait for all Controller traits.
+pub trait Controller<Origin, RuntimeCall, Timeout>:
+	ExecuteController<Origin, RuntimeCall> + SendController<Origin> + QueryController<Origin, Timeout>
+{
+}
+
+impl<T, Origin, RuntimeCall, Timeout> Controller<Origin, RuntimeCall, Timeout> for T where
+	T: ExecuteController<Origin, RuntimeCall>
+		+ SendController<Origin>
+		+ QueryController<Origin, Timeout>
+{
+}
+
+/// Weight functions needed for [`ExecuteController`].
+pub trait ExecuteControllerWeightInfo {
+	/// Weight for [`ExecuteController::execute`]
+	fn execute() -> Weight;
+}
+
+/// Execute an XCM locally, for a given origin.
+pub trait ExecuteController<Origin, RuntimeCall> {
+	/// Weight information for ExecuteController functions.
+	type WeightInfo: ExecuteControllerWeightInfo;
+
+	/// Attempt to execute an XCM locally, and return the outcome.
+	///
+	/// # Parameters
+	///
+	/// - `origin`: the origin of the call.
+	/// - `message`: the XCM program to be executed.
+	/// - `max_weight`: the maximum weight that can be consumed by the execution.
+	fn execute(
+		origin: Origin,
+		message: Box<VersionedXcm<RuntimeCall>>,
+		max_weight: Weight,
+	) -> Result<Outcome, DispatchError>;
+}
+
+/// Weight functions needed for [`SendController`].
+pub trait SendControllerWeightInfo {
+	/// Weight for [`SendController::send`]
+	fn send() -> Weight;
+}
+
+/// Send an XCM from a given origin.
+pub trait SendController<Origin> {
+	/// Weight information for SendController functions.
+	type WeightInfo: SendControllerWeightInfo;
+
+	/// Send an XCM to be executed by a remote location.
+	///
+	/// # Parameters
+	///
+	/// - `origin`: the origin of the call.
+	/// - `dest`: the destination of the message.
+	/// - `msg`: the XCM to be sent.
+	fn send(
+		origin: Origin,
+		dest: Box<VersionedMultiLocation>,
+		message: Box<VersionedXcm<()>>,
+	) -> Result<XcmHash, DispatchError>;
+}
+
+/// Weight functions needed for [`QueryController`].
+pub trait QueryControllerWeightInfo {
+	/// Weight for [`QueryController::query`]
+	fn query() -> Weight;
+
+	/// Weight for [`QueryController::take_response`]
+	fn take_response() -> Weight;
+}
+
+/// Query a remote location, from a given origin.
+pub trait QueryController<Origin, Timeout>: QueryHandler {
+	/// Weight information for QueryController functions.
+	type WeightInfo: QueryControllerWeightInfo;
+
+	/// Query a remote location.
```

Comment:
What does this do? Should it maybe be in the `QueryHandler` trait we already have? Same question with the weight info

Reply:
in short what the other controller traits are doing, dealing with low level details (here converting and versioning) and calling QueryHandler api to register a new query.

https://github.com/paritytech/polkadot-sdk//blob/634296291c56d38bfd6c7903d490465d110f98ef/polkadot/xcm/pallet-xcm/src/lib.rs#L354-L373